Classes/git/PBGitIndex.m

```diff
@@ -521,8 +521,24 @@ - (BOOL)applyPatch:(NSString *)hunk stage:(BOOL)stage reverse:(BOOL)reverse;
 		return NO;
 	}
 
-	// TODO: Try to be smarter about what to refresh
-	[self refresh];
+	// Get the status for the updated file and update the index
+	BOOL success = NO;
+	GTFileStatusFlags status = [[[self repository] gtRepo] statusForFile:[file path] success:&success error:NULL];
+
+	if (success) {
+		// Search for file in [self files]
+		// file is probably a member of [self files], but let's not assume that to be safe
+		for (PBChangedFile *nextFile in [self files]) {
+			if (nextFile == file || [[nextFile path] isEqualToString:[file path]]) {
+				[nextFile setHasUnstagedChanges:(status & GTFileStatusModifiedInWorktree) != 0];
+				[nextFile setHasStagedChanges:(status & GTFileStatusModifiedInIndex) != 0];
+
+				[self postIndexUpdated];
```

Comment:
In trying to understand what "Fixes a UI problem where staging the last line or last chunk of a modified file doesn't remove the file from the list of unstaged files.", I came to the conclusion that this is the call responsible for it. But I'm not sure at all this is the case…

Reply:
Here's how to reproduce the problem.

1. Modify a single line in a file so it appears in unstaged changes
2. Select the line in the diff view and stage just that line
3. This results in the entire file being stated, as there's no remaining unstaged changes
4. gitx doesn't update its state, leaving hasUnstagedChanges` set to YES until a manual index refresh (so the file appears in both unstaged and staged changes, even though it should appear only in staged changes)

This skips calling refresh when staging an individual patch, as we know we're only changing one file. This means we can update the status of that one file rather than updating the entire index. This both avoids the problem and results in a speedup because we're checking the status of one file rather than everything (plus it's using libgit2 instead of going through the git command line).